vllm/usage/usage_lib.py

```diff
@@ -0,0 +1,139 @@
+    return _USAGE_STATS_ENABLED
+
+
+def _get_current_timestamp_ns() -> int:
+    return int(datetime.datetime.now(datetime.timezone.utc).timestamp() * 1e9)
+
+
+def _detect_cloud_provider() -> str:
+    # Try detecting through vendor file
+    vendor_files = [
+        '/sys/class/dmi/id/product_version', '/sys/class/dmi/id/bios_vendor',
+        '/sys/class/dmi/id/product_name',
+        '/sys/class/dmi/id/chassis_asset_tag', '/sys/class/dmi/id/sys_vendor'
+    ]
+    # Mapping of identifiable strings to cloud providers
+    cloud_identifiers = {
+        'amazon': "AWS",
+        'microsoft corporation': "AZURE",
+        'google': "GCP",
+        'oraclecloud': "OCI",
+    }
+
+    for vendor_file in vendor_files:
+        path = Path(vendor_file)
+        if path.is_file():
+            file_content = path.read_text().lower()
+            for identifier, provider in cloud_identifiers.items():
+                if identifier in file_content:
+                    return provider
+    return "UNKNOWN"
+
+
+class UsageContext(Enum):
+    UNKNOWN_CONTEXT = "UNKNOWN_CONTEXT"
+    LLM_CLASS = "LLM_CLASS"
+    API_SERVER = "API_SERVER"
+    OPENAI_API_SERVER = "OPENAI_API_SERVER"
+    ENGINE_CONTEXT = "ENGINE_CONTEXT"
+
+
+class UsageMessage:
+
+    def __init__(self) -> None:
+        self.gpu_list: Optional[dict] = None
+        self.provider: Optional[str] = None
+        self.architecture: Optional[str] = None
+        self.platform: Optional[str] = None
+        self.model: Optional[str] = None
+        self.vllm_version: Optional[str] = None
+        self.context: Optional[str] = None
+        self.log_time: Optional[int] = None
+        #Logical CPU count
+        self.num_cpu: Optional[int] = None
+        self.cpu_type: Optional[str] = None
+        self.total_memory: Optional[int] = None
+        self.source: Optional[str] = None
+
+    def report_usage(self, model: str, context: UsageContext) -> None:
+        t = Thread(target=usage_message._report_usage, args=(model, context))
+        t.start()
+
+    def _report_usage(self, model: str, context: UsageContext) -> None:
+        self.context = context.value
+        self.gpu_list = []
+        for i in range(torch.cuda.device_count()):
+            device_property = torch.cuda.get_device_properties(i)
+            gpu_name = device_property.name
+            gpu_memory = device_property.total_memory
+            self.gpu_list.append({"name": gpu_name, "memory": gpu_memory})
+        self.provider = _detect_cloud_provider()
+        self.architecture = platform.machine()
+        self.platform = platform.platform()
+        self.vllm_version = pkg_resources.get_distribution("vllm").version
+        self.model = model
+        self.log_time = _get_current_timestamp_ns()
+        self.num_cpu = os.cpu_count()
+        #Best effort reading processor name
+        self.cpu_type = platform.processor()
+        self.total_memory = psutil.virtual_memory().total
```

Comment:
Just a heads up - if the model server is deployed as a linux docker container, then most metrics from `os` and `psutil` will reflect the overall capacity of the machine instead of the actual resources allocated to the container. This has been an issue for a long time - see https://github.com/python/cpython/issues/80235

Reply:
Yeah i know there's a way around this but a bit too complex. i'm just going to assume most folks who are running LLM prod will adopt a one-pod-per-vm approach